# SPERO LAB TnSeq (Barcode-Mapping) Pipeline

This repository contains the scripts and materials required to run the Spero Lab's TnSeq Barcode-Mapping Pipeline. 

Required input for this pipeline include 1. FASTQ files from samples, 2. a FASTA file for reference genome, and 3. a GTF file for reference genome. 

Once succesfully run, the pipeline should output a TSV file mapping each unique tranposon barcode to the GeneID of its targetted genomic loci. Additionally, a R markdown report discussing overall read quality as well as insertion saturation density will be output into the results folder. 

The repository for this pipeline is organized as such:

TnSeq_Barcode_Map/
- config/ → contains .yml files where users can specify adjustable parameters that influence pipeline output
- data/ → designated folder to put all input file before running pipeline
- results/ → stores output files generated by pipeline, including processed data for downstream analyses and quality control reports 
- scripts/ → includes pipeline scripts (code) that perform data processing and analysis steps — modifying these files may disrupt pipeline functionality

For more details on the functionality and details of this pipeline, please refer to the manual (nternship Report #1) submitted by Kenneth Lai or contact them by kenneth.lok.sun.lai@gmail.com. 



